spec/index.md

```diff
@@ -1550,6 +1564,14 @@ If this traps (e.g. runs out of cycles), the state changes from the `cleanup` fu
 
 There must be at most one call to `ic0.call_on_cleanup` between `ic0.call_new` and `ic0.call_perform`.
 
+-   `ic0.call_with_best_effort_response : (timeout_seconds : i32) -> ()`
+
+    Allows the system to generate a synthetic reject response without waiting for the actual response from the callee, even if the callee produces a response. The reject callback will be triggered with either `SYS_TRANSIENT` or `SYS_UNKNOWN` reject codes. If the code is `SYS_TRANSIENT`, the caller can be sure that the call never made it to the callee. If it is `SYS_UNKNOWN`, then the call **may or may not** have been processed by the callee; for example, it could be that processing the call took too long, that the reply took too long to deliver to the calling canister, or simply that the system decided that it needed to shed load. In this case, if the calling canister needs to know whether the call was successful, it must find an out-of-band way of doing so. For example, if the callee provides idempotent function calls, the calling canister can simply retry the call.
+
+    The `timeout_seconds` parameter is a hint to the system when the reject should be produced. However, it is only advisory. First, it is silently bounded by the `MAX_CALL_TIMEOUT` system constant; i.e., larger timeouts are treated as equivalent to `MAX_CALL_TIMEOUT` and do not cause an error. Second, the system may invoke the reject callback earlier than requested (e.g., under high load); the calling canister should not rely on the timeout not occurring early. Finally, the reject callback may also be executed (possibly significantly) later than the specified time, e.g., if the canister is under high load.
```

Comment:
Personal opinion: I would separate timing out a call from dropping it to shed load. That way a timeout reject would only ever be produced after (when?) the deadline has expired.

But similar to how we can already reject a request with a `SYS_TRANSIENT` reject code at any time, because the caller explicitly selected to make a best-effort call, the system could also terminate the caller's end of the call at any time with a `SYS_UNKNOWN` reject code. Without it involving a time out.

Reply:
Unless we want to define another reject code, I don't see how we can't distinguish between timeouts and premature dropping at the interface spec level, given the current choice not to put ICXXX codes that aren't specified in the interface spec?

Reply:
We should definitely not define another reject code.

Since no IC*XXX* error codes are defined in the spec, there is no documented way of figuring out e.g. what kind of transient error has occurred (queue full, out of memory, subnet split, whatever). By that measure, we shouldn't concern ourselves with how to differentiate between the various `SYS_UNKNOWN` root causes (for now only "timed out" and "dropped").

All we need to do is state that there are 2 possible reasons why a best-effort call might end up in an unknown state: timeouts and load shedding. Bundling both under timeouts -- and having to explain that load shedding is essentially an early timeout and thus your specified timeout is meaningless -- is backwards IMHO. No other reject code has a single root cause behind it, so we shouldn't aim to make `SYS_UNKNOWN` different.

Reply:
There's https://wiki.internetcomputer.org/wiki/Error_Codes_returned_by_Internet_Computer which documents the embedded error codes, we should add them here.

I've removed "timeout too early now", I hope that helps. Please suggest a rewording if you still think it's confusing. My ask is just that we clearly spell out to the spec user that the reject callback can be triggered both earlier and later than the specified timeout.

Reply:
```suggestion
    The `timeout_seconds` parameter specifies how long the system should wait for a response before producing a timeout reject. It is silently bounded by the `MAX_CALL_TIMEOUT` system constant; i.e., larger timeouts are treated as equivalent to `MAX_CALL_TIMEOUT` and do not cause an error. The reject callback may be executed (possibly significantly) later than the specified time, e.g., if the canister is under high load. Also note that the system may reject a call for other reasons (e.g., under high load); the calling canister should not rely on the reject callback not being called before timeout expiration.
```

Maybe also squeeze in something about `SYS_UNKNOWN` and `SYS_TRANSIENT`. Both timeouts and other rejects may produce either reject code.

Reply:
Yes I realize that the first paragraph was wrong in that respect 😬 I modified it again now...